fem/bilininteg_ext.cpp

```diff
@@ -0,0 +1,1588 @@
+   const FiniteElementSpace *fespace = nodes->FESpace();
+   const FiniteElement *fe = fespace->GetFE(0);
+   const IntegrationRule& ir1D = IntRules.Get(Geometry::SEGMENT,ir.GetOrder());
+   const int dims     = fe->GetDim();
+   const int numDofs  = fe->GetDof();
+   const int D1D      = fe->GetOrder() + 1;
+   const int Q1D      = ir1D.GetNPoints();
+   const int elements = fespace->GetNE();
+   const int ndofs    = fespace->GetNDofs();
+   const DofToQuad* maps = DofToQuad::GetSimplexMaps(*fe, ir);
+   NodeCopyByVDim(elements,numDofs,ndofs,dims,geom->eMap,Sx,geom->nodes);
+   PAGeom(dims, D1D, Q1D, elements,
+          maps->B, maps->G, geom->nodes,
+          geom->X, geom->J, geom->invJ, geom->detJ);
+   return geom;
+}
+
+GeometryExtension* GeometryExtension::Get(const FiniteElementSpace& fes,
+                                          const IntegrationRule& ir)
+{
+   Mesh *mesh = fes.GetMesh();
+   const bool geom_to_allocate = sequence < fes.GetSequence();
+   sequence = fes.GetSequence();
+   if (geom_to_allocate) { geom = new GeometryExtension(); }
+   mesh->EnsureNodes();
+   const GridFunction *nodes = mesh->GetNodes();
+   const mfem::FiniteElementSpace *fespace = nodes->FESpace();
+   const mfem::FiniteElement *fe = fespace->GetFE(0);
+   const IntegrationRule& ir1D = IntRules.Get(Geometry::SEGMENT,ir.GetOrder());
+   const int dims     = fe->GetDim();
+   const int elements = fespace->GetNE();
+   const int numDofs  = fe->GetDof();
+   const int D1D      = fe->GetOrder() + 1;
+   const int Q1D      = ir1D.GetNPoints();
+   const int numQuad  = ir.GetNPoints();
+   const bool orderedByNODES = (fespace->GetOrdering() == Ordering::byNODES);
+   if (orderedByNODES) { ReorderByVDim(nodes); }
+   const int asize = dims*numDofs*elements;
+   mfem::Array<double> meshNodes(asize);
+   const Table& e2dTable = fespace->GetElementToDofTable();
+   const int* elementMap = e2dTable.GetJ();
+   mfem::Array<int> eMap(numDofs*elements);
+   GeomFill(dims,
+            elements,
+            numDofs,
+            nodes->Size(),
+            elementMap,
+            eMap,
+            nodes->GetData(),
+            meshNodes);
+   if (geom_to_allocate)
+   {
+      geom->nodes.SetSize(dims*numDofs*elements);
+      geom->eMap.SetSize(numDofs*elements);
+   }
+   geom->nodes = meshNodes;
+   geom->eMap = eMap;
+   // Reorder the original gf back
+   if (orderedByNODES) { ReorderByNodes(nodes); }
+   if (geom_to_allocate)
+   {
+      geom->X.SetSize(dims*numQuad*elements);
+      geom->J.SetSize(dims*dims*numQuad*elements);
+      geom->invJ.SetSize(dims*dims*numQuad*elements);
+      geom->detJ.SetSize(numQuad*elements);
+   }
+   const DofToQuad* maps = DofToQuad::GetSimplexMaps(*fe, ir);
+   PAGeom(dims, D1D, Q1D, elements,
+          maps->B, maps->G, geom->nodes,
+          geom->X, geom->J, geom->invJ, geom->detJ);
+   return geom;
+}
+
+void GeometryExtension::ReorderByVDim(const GridFunction *nodes)
```

Comment:
Instead of reordering, I think it would be better to have algorithm that take into account the ordering in the future, since this reordering is potentially expensive I think.